go/genkit/flow.go

```diff
@@ -81,48 +85,70 @@ import (
 // no way to schedule or resume a flow with streaming.
 
 // A Flow is an Action with additional support for observability and introspection.
 // A Flow[In, Out, Stream] represents a function from In to Out. The Stream parameter is for
 // flows that support streaming: providing their results incrementally.
 type Flow[In, Out, Stream any] struct {
-	name         string                // The last component of the flow's key in the registry.
-	fn           Func[In, Out, Stream] // The function to run.
-	stateStore   FlowStateStore        // Where FlowStates are stored, to support resumption.
-	tstate       *tracing.State        // set from the action when the flow is defined
-	inputSchema  *jsonschema.Schema    // Schema of the input to the flow
-	outputSchema *jsonschema.Schema    // Schema of the output out of the flow
+	name         string                     // The last component of the flow's key in the registry.
+	fn           core.Func[In, Out, Stream] // The function to run.
+	stateStore   core.FlowStateStore        // Where FlowStates are stored, to support resumption.
+	tstate       *tracing.State             // set from the action when the flow is defined
+	inputSchema  *jsonschema.Schema         // Schema of the input to the flow
+	outputSchema *jsonschema.Schema         // Schema of the output out of the flow
 	// TODO(jba): scheduler
 	// TODO(jba): experimentalDurable
 	// TODO(jba): authPolicy
 	// TODO(jba): middleware
 }
 
-// InternalDefineFlow is for use by genkit.DefineFlow exclusively.
-// It is not subject to any backwards compatibility guarantees.
-func InternalDefineFlow[In, Out, Stream any](name string, fn Func[In, Out, Stream]) *Flow[In, Out, Stream] {
-	return defineFlow(globalRegistry, name, fn)
-}
+type noStream = func(context.Context, struct{}) error
 
-func defineFlow[In, Out, Stream any](r *registry, name string, fn Func[In, Out, Stream]) *Flow[In, Out, Stream] {
+// DefineFlow creates a Flow that runs fn, and registers it as an action.
+//
+// fn takes an input of type In and returns an output of type Out.
+func DefineFlow[In, Out any](
+	name string,
+	fn func(ctx context.Context, input In) (Out, error),
+) *Flow[In, Out, struct{}] {
+	return defineFlow(registry.Global, name, core.Func[In, Out, struct{}](
+		func(ctx context.Context, input In, cb func(ctx context.Context, _ struct{}) error) (Out, error) {
+			return fn(ctx, input)
+		}))
+}
+
+// DefineStreamingFlow creates a streaming Flow that runs fn, and registers it as an action.
+//
+// fn takes an input of type In and returns an output of type Out, optionally
+// streaming values of type Stream incrementally by invoking a callback.
+//
+// If the function supports streaming and the callback is non-nil, it should
+// stream the results by invoking the callback periodically, ultimately returning
+// with a final return value. Otherwise, it should ignore the callback and
+// just return a result.
+func DefineStreamingFlow[In, Out, Stream any](
```

Comment:
Might want to mention that we expect all the data to be available from the final result. Streaming results are just a preview; the final result object contains (at least) all the streaming responses.

Reply:
Done.